x/merkledb/view.go

```diff
@@ -290,12 +290,63 @@ func (v *view) hashChangedNodes(ctx context.Context) {
 // Calculates the ID of all descendants of [n] which need to be recalculated,
 // and then calculates the ID of [n] itself.
 func (v *view) hashChangedNode(n *node) ids.ID {
-	// We use [wg] to wait until all descendants of [n] have been updated.
-	var wg sync.WaitGroup
+	// If there are no children, we can avoid allocating [keyBuffer].
+	if len(n.children) == 0 {
+		return n.calculateID(v.db.metrics)
+	}
+
+	// Calculate the size of the largest child key of this node. This allows
+	// only allocating a single slice for all of the keys.
+	var maxBitLength int
+	for _, childEntry := range n.children {
+		totalBitLength := n.key.length + v.tokenSize + childEntry.compressedKey.length
+		maxBitLength = max(maxBitLength, totalBitLength)
+	}
+
+	var (
+		maxBytesNeeded = bytesNeeded(maxBitLength)
+		// keyBuffer is allocated onto the heap because it is dynamically sized.
+		keyBuffer = make([]byte, maxBytesNeeded)
+		// childBuffer is allocated on the stack.
+		childBuffer = make([]byte, 1)
+		dualIndex   = dualBitIndex(v.tokenSize)
+		bytesForKey = bytesNeeded(n.key.length)
+		// We track the last byte used by the key so that we can reset the
+		// value. for each key. This is needed because the child buffer may get
+		// ORed with this byte.
+		lastKeyByte byte
+
+		// We use [wg] to wait until all descendants of [n] have been updated.
+		wg sync.WaitGroup
+	)
+
+	if bytesForKey > 0 {
+		// We can just copy this node's key once. It doesn't change as we
+		// iterate over the children.
+		copy(keyBuffer, n.key.value)
+		lastKeyByte = keyBuffer[bytesForKey-1]
+	}
 
 	for childIndex, childEntry := range n.children {
-		childEntry := childEntry // New variable so goroutine doesn't capture loop variable.
-		childKey := n.key.Extend(ToToken(childIndex, v.tokenSize), childEntry.compressedKey)
+		childBuffer[0] = childIndex << dualIndex
+		childByteKey := Key{
```

Comment:
nit: I think this is slightly more clear but I'm ok with it as is too
```suggestion
		childIndexAsKey := Key{
```